### PR TITLE
script for doing archive surgery

### PR DESCRIPTION
**Pull request by joelgrus**
allows you to modify a model.tar.gz by either 

* inspecting its config
* adding a new key-value pair
* deleting an existing key-value pair
* moving a value to a different location (which includes renaming a key)
* replacing a value

unfortunately, it can only do one of these at a time, I think that covers a lot of use cases, but the usage would get a lot more complicated if I allowed multiple operations in one call.

making the `files_to_archive` piece work right was sort of intricate

--

the motivating cause was having to update the trained quarel model, which itself depended on a trained parser model, which was using a deprecated key in its config.

the entire workflow was as follows:

1. download the model

```
wget https://s3-us-west-2.amazonaws.com/allennlp/models/quarel-parser-zero-2018.10.03.tar.gz
```

2. inspect its config

```
scripts/archive_surgery.py inspect --input-file quarel-parser-zero-2018.10.03.tar.gz
```

inspection reveals that it depends on a tagger model in two places:

`validation_dataset_reader.world_extraction_model: https://s3-us-west-2.amazonaws.com/allennlp/models/quarel-tagger-2018.10.03.tar.gz`
`dataset_reader.world_extraction_model: https://s3-us-west-2.amazonaws.com/allennlp/models/quarel-tagger-2018.10.03.tar.gz`

so

3. download that tagger model

```
wget https://s3-us-west-2.amazonaws.com/allennlp/models/quarel-tagger-2018.10.03.tar.gz
```

4. inspect it

```
scripts/archive_surgery.py inspect --input-file quarel-tagger-2018.10.03.tar.gz
```

it has a deprecated entry: `model.constraint_type: BIO`, which we need to 

5. rename to `model.label_encoding`

```
scripts/archive_surgery.py move --input-file quarel-tagger-2018.10.03.tar.gz --output-file quarel-tagger-2018.12.20.tar.gz --old-key model.constraint_type --new-key model.label_encoding
```

and then

6. upload to s3

```
aws s3 cp quarel-tagger-2018.12.20.tar.gz s3://allennlp/models/quarel-tagger-2018.12.20.tar.gz --acl public-read
```

now we need to replace the two values that referred to the old tagger model. unfortunately, we need to do that in two steps, because the surgery script doesn't handle multiple changes.

7. replace one, save to intermediate file:

```
scripts/archive_surgery.py replace --input-file quarel-parser-zero-2018.10.03.tar.gz --output-file quarel-parser-zero-2018.12.20.tar.gz.intermediate --key validation_dataset_reader.world_extraction_model --value https://s3-us-west-2.amazonaws.com/allennlp/models/quarel-tagger-2018.12.20.tar.gz
```

8. replace the other, save to final file

```
scripts/archive_surgery.py replace --input-file quarel-parser-zero-2018.12.20.tar.gz.intermediate --output-file quarel-parser-zero-2018.12.20.tar.gz --key dataset_reader.world_extraction_model --value https://s3-us-west-2.amazonaws.com/allennlp/models/quarel-tagger-2018.12.20.tar.gz
```

9. finally, upload the fixed model to s3

```
aws s3 cp quarel-parser-zero-2018.12.20.tar.gz s3://allennlp/models/quarel-parser-zero-2018.12.20.tar.gz --acl public-read
```

### CODE REVIEW

File: scripts/archive_surgery.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    remove_parser.set_defaults(func=remove)
+
+    inspect_parser.add_argument("--input-file", required=True)
+    inspect_parser.set_defaults(func=inspect)
+
+    args = parser.parse_args()
+    command = args.command
+
+    if command != "inspect" and os.path.exists(args.output_file):
+        raise ValueError("output file already exists")
```
Comment (schmmd) [inline scripts/archive_surgery.py]: I wonder if you get this for free with `argparse.FileType('w')`

File: scripts/archive_surgery.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    original_config = Params.from_file(config_path, overrides)
+
+    # Do surgery
+    new_config: Params = args.func(args, original_config)
+
+    # If we're just inspecting, we can quit now
+    if command == "inspect":
+        sys.exit(0)
+
+    # Otherwise let's print out the new config
```
Comment (schmmd) [inline scripts/archive_surgery.py]: "...to stdout"

**Comment by schmmd** [comment]
If you want to be able to chain, you could rely more on stdin/stdout, e.g.

```
cat quarel-parser-zero-2018.10.03.tar.gz | scripts/archive_surgery.py replace --key validation_dataset_reader.world_extraction_model --value https://s3-us-west-2.amazonaws.com/allennlp/models/quarel-tagger-2018.12.20.tar.gz |
scripts/archive_surgery.py replace --key dataset_reader.world_extraction_model --value https://s3-us-west-2.amazonaws.com/allennlp/models/quarel-tagger-2018.12.20.tar.gz > quarel-parser-zero-2018.12.20.tar.gz 
```

**Comment by matt-gardner** [comment]
I'm not sure how feasible this is, but would things be simpler if you had the script pop out to `EDITOR`?

**Comment by joelgrus** [comment]
I'm really leery of writing .tar.gz files to stdout by default. (And each step unzips and rezips everything, so if people frequently need multiple changes, I'd rather design to do it in one step, it's just hard to get the API right for that and I wanted to get this done.)

**Comment by schmmd** [comment]
@matt-gardner my approach to updating these tar.gzs is to simply use vim on the tarball directly, which works well for me.

**Comment by matt-peters** [comment]
Since there is a quick, easy understand, and easy to describe method for accomplishing this (untar the archive, edit the config, re-tar) is it really necessary to have the 200+ line script to accomplish the same thing?  A line of code not written is worth two lines written? Or something like that.

**Comment by joelgrus** [comment]
@matt-peters my main motivation was that every time we deprecate a config key, we *break* every trained model.tar.gz that exists in the world, and I wanted to make it as easy as possible for our users to fix their configs.

that is, I wanted to be able to tell them "run this one line command" and it will fix your model, rather than "untar it, find the config, open a text editor, make changes, then re-tar it", which is fine if you're used to doing it but not a great user experience.

but if people think this is not a great idea, I'm not wedded to it

**Comment by nelson-liu** [comment]
fwiw: i just tried to modify a config in one of the fixtures, and the tests weren't taking the modified config (archive loading code complains about `tarfile.ReadError: not a gzip file`). Using this script worked beautifully, thanks.

**Comment by schmmd** [comment]
Seems like a significant code reduction ;-)

**Pull request merged**